include/pisa/binary_collection.hpp

```diff
@@ -11,131 +11,129 @@
-                // file might be truncated
-                n = std::min(n, size_t(m_collection->m_data_size - pos));
-                auto begin = &m_collection->m_data[pos];
-
-                m_next_pos = pos + n;
-                m_cur_seq  = S(begin, begin + n);
-            }
-
-            base_binary_collection const *     m_collection;
-            size_t                             m_pos, m_next_pos;
-            S                                  m_cur_seq;
-        };
+       private:
+        pointer m_begin;
+        pointer m_end;
+    };
+
+    using const_sequence = sequence;
+
+    template <typename S>
+    class base_iterator;
+
+    using const_iterator = base_iterator<const_sequence>;
+    using iterator       = typename std::conditional<std::is_same<Source, mio::mmap_source>::value,
+                                               const_iterator, base_iterator<sequence>>::type;
+
+    iterator       begin() { return iterator(this, 0); }
+    iterator       end() { return iterator(this, m_data_size); }
+    const_iterator begin() const { return const_iterator(this, 0); }
+    const_iterator end() const { return const_iterator(this, m_data_size); }
+    const_iterator cbegin() const { return const_iterator(this, 0); }
+    const_iterator cend() const { return const_iterator(this, m_data_size); }
+
+    template <typename S>
+    class base_iterator : public std::iterator<std::forward_iterator_tag, S> {
+       public:
+        base_iterator() : m_data(nullptr) {}
+
+        auto const &operator*() const { return m_cur_seq; }
+
+        auto const *operator-> () const { return &m_cur_seq; }
+
+        base_iterator &operator++() {
+            m_pos = m_next_pos;
+            read();
+            return *this;
+        }
+
+        bool operator==(base_iterator const &other) const {
+            assert(m_data == other.m_data);
+            assert(m_data_size == other.m_data_size);
+            return m_pos == other.m_pos;
+        }
+
+        bool operator!=(base_iterator const &other) const { return !(*this == other); }
 
        private:
-        Source  m_file;
-        pointer m_data;
-        size_t  m_data_size;
+        friend class base_binary_collection;
+
+        base_iterator(base_binary_collection const *coll, size_t pos)
+            : m_data(coll->m_data), m_data_size(coll->m_data_size), m_pos(pos) {
+            read();
+        }
+
+        void read() {
+            assert(m_pos <= m_data_size);
+            if (m_pos == m_data_size) return;
+
+            size_t n   = 0;
+            size_t pos = m_pos;
+            n          = m_data[pos++];
+            // file might be truncated
+            n          = std::min(n, size_t(m_data_size - pos));
+            auto begin = &m_data[pos];
+
+            m_next_pos = pos + n;
+            m_cur_seq  = S(begin, begin + n);
+        }
+
+        const pointer m_data;
+        size_t        m_data_size = 0, m_pos = 0, m_next_pos = 0;
```

Comment:
I would avoid this in the future, it's inconsistent where to look for variable names: if we have a set of lines, then let's stick to it.